File: test/assess.test.ts
// Post-synthesis confidence gate — gating what may be *acted on*.
//
// The planner decides what may be loaded; grounding decides what may be
// asserted; assess() decides whether a conclusion clears the caller's
// confidence threshold before it is acted on. Same trust posture as
// grounding: confidence is a proposal (self-report or an injected
// evaluator); the gate adjudicates deterministically.
//
// Written test-first. The verdict reuses the GateVerdict contract
// (gate/passed/detail) but is a downstream artifact — it is never
// inserted into DecisionTrace, so conformance vectors are untouched.

import { describe, it, expect } from "vitest";
import {
  assess,
  extractSelfReport,
  makeProviderEvaluator,
  type ConfidenceEvaluator,
  type ConfidenceSignal,
} from "../src/assess.js";
import type { GroundUnit } from "../src/ground.js";
import type { SynthesisProvider } from "../src/provider.js";

const U = (id: string, content: string): GroundUnit => ({ id, sha256: `sha-${id}`, content });
const UNITS = [U("risk-policy", "Risk assessments require dual sign-off.")];

const signal = (source: "self" | "evaluator", score: number, reasoning = "because"): ConfidenceSignal =>
  ({ source, score, reasoning });

const evaluatorOf = (score: number, reasoning = "evaluator judgment"): ConfidenceEvaluator =>
  async () => signal("evaluator", score, reasoning);

describe("assess", () => {
  it("passes when the adjudicated score clears the threshold", async () => {
    const v = await assess("draft risk assessment", "Low risk.", UNITS, {
      threshold: 0.7,
      selfReport: signal("self", 0.9, "clear-cut case"),
    });
    expect(v.gate).toBe("confidence");
    expect(v.passed).toBe(true);
    expect(v.score).toBe(0.9);
    expect(v.threshold).toBe(0.7);
    expect(v.signals).toHaveLength(1);
    expect(v.detail).toContain("0.9");
    expect(v.detail).toContain("0.7");
  });

  it("fails with a written, specific reason when below threshold", async () => {
    const v = await assess("draft risk assessment", "Unsure.", UNITS, {
      threshold: 0.7,
      severity: "critical",
      selfReport: signal("self", 0.55, "conflicting inputs"),
    });
    expect(v.passed).toBe(false);
    expect(v.score).toBe(0.55);
    expect(v.severity).toBe("critical");
    expect(v.detail).toContain("0.55");
    expect(v.detail).toContain("conflicting inputs");
  });

  it("aggregates multiple signals with min by default — fail-closed", async () => {
    const v = await assess("task", "Answer. Confidence: 0.9", UNITS, {
      threshold: 0.7,
      evaluator: evaluatorOf(0.6, "citations are thin"),
    });
    expect(v.signals).toHaveLength(2);
    expect(v.score).toBe(0.6);
    expect(v.passed).toBe(false);
    expect(v.detail).toContain("citations are thin");
  });

  it("aggregate: mean averages the signals", async () => {
    const v = await assess("task", "Answer. Confidence: 0.9", UNITS, {
      threshold: 0.7,
      evaluator: evaluatorOf(0.6),
      aggregate: "mean",
    });
    expect(v.score).toBeCloseTo(0.75);
    expect(v.passed).toBe(true);
  });

  it("no obtainable signal → fail-closed with a specific detail", async () => {
    const v = await assess("task", "An answer with no self-report.", UNITS, { threshold: 0.7 });
    expect(v.passed).toBe(false);
    expect(v.score).toBe(0);
    expect(v.signals).toHaveLength(0);
    expect(v.detail).toMatch(/no confidence signal/i);
  });

  it("evaluator failure → fail-closed, the error becomes the detail", async () => {
    const broken: ConfidenceEvaluator = async () => { throw new Error("provider offline"); };
    const v = await assess("task", "Answer. Confidence: 0.95", UNITS, {
      threshold: 0.7,
      evaluator: broken,
    });
    expect(v.passed).toBe(false);
    expect(v.detail).toContain("provider offline");
  });

  it("evaluator returning an out-of-range score → fail-closed", async () => {
    const v = await assess("task", "Answer.", UNITS, {
      threshold: 0.7,
      evaluator: evaluatorOf(42, "not calibrated"),
    });
    expect(v.passed).toBe(false);
    expect(v.detail).toMatch(/out of range|invalid/i);
  });

  it("raw signals are preserved verbatim for calibration", async () => {
    const v = await assess("task", "Answer. Confidence: 80%", UNITS, {
      threshold: 0.5,
      evaluator: evaluatorOf(0.9, "well grounded"),
    });
    const sources = v.signals.map((s) => s.source).sort();
    expect(sources).toEqual(["evaluator", "self"]);
    for (const s of v.signals) {
      expect(s.reasoning).toBeTruthy();
    }
  });

  it("includeSelfReport: false ignores the answer's self-report", async () => {
    const v = await assess("task", "Answer. Confidence: 0.2", UNITS, {
      threshold: 0.7,
      evaluator: evaluatorOf(0.9),
      includeSelfReport: false,
    });
    expect(v.signals).toHaveLength(1);
    expect(v.score).toBe(0.9);
    expect(v.passed).toBe(true);
  });

  it("stamps asOf (caller-provided wins, for reproducibility)", async () => {
    const v = await assess("task", "Answer.", UNITS, {
      threshold: 0.7,
      selfReport: signal("self", 0.8),
      asOf: "2026-07-20",
    });
    expect(v.asOf).toBe("2026-07-20");
  });

  it("rejects an invalid threshold — caller error, not a verdict", async () => {
    await expect(assess("task", "Answer.", UNITS, { threshold: 7 })).rejects.toThrow(/threshold/);
    await expect(assess("task", "Answer.", UNITS, { threshold: -0.1 })).rejects.toThrow(/threshold/);
  });
});

describe("extractSelfReport", () => {
  it("parses a decimal confidence line", () => {
    const s = extractSelfReport("The risk is low.\n\nConfidence: 0.82");
    expect(s?.score).toBeCloseTo(0.82);
    expect(s?.source).toBe("self");
  });

  it("parses a percentage", () => {
    expect(extractSelfReport("Confidence: 82%")?.score).toBeCloseTo(0.82);
  });

  it("takes the last report when several appear", () => {
    const s = extractSelfReport("Confidence: 0.9 early on.\nFinal confidence: 0.6");
    expect(s?.score).toBeCloseTo(0.6);
  });

  it("returns undefined when no self-report exists", () => {
    expect(extractSelfReport("Just an answer.")).toBeUndefined();
  });

  it("clamps runaway percentages into range", () => {
    expect(extractSelfReport("Confidence: 150%")?.score).toBe(1);
  });
});

describe("makeProviderEvaluator", () => {
  const providerReturning = (text: string): SynthesisProvider => ({
    name: "fake",
    complete: async () => text,
    stream: async function* () { yield text; },
  }) as unknown as SynthesisProvider;

  it("parses the provider's JSON verdict into a signal", async () => {
    const ev = makeProviderEvaluator(providerReturning('{"score": 0.85, "reasoning": "claims match units"}'));
    const s = await ev({ task: "t", answer: "a", units: UNITS });
    expect(s.source).toBe("evaluator");
    expect(s.score).toBeCloseTo(0.85);
    expect(s.reasoning).toBe("claims match units");
  });

  it("strips markdown fences", async () => {
    const ev = makeProviderEvaluator(providerReturning('```json\n{"score": 0.5, "reasoning": "mixed"}\n```'));
    expect((await ev({ task: "t", answer: "a", units: UNITS })).score).toBe(0.5);
  });

  it("unparseable verdict → score 0, fail-closed", async () => {
    const ev = makeProviderEvaluator(providerReturning("I feel pretty good about it"));
    const s = await ev({ task: "t", answer: "a", units: UNITS });
    expect(s.score).toBe(0);
    expect(s.reasoning).toMatch(/unparseable/i);
  });
});

File: java/kcp-planner/src/test/java/no/cantara/kcp/planner/assess/AssessTest.java
package no.cantara.kcp.planner.assess;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;

import org.junit.jupiter.api.Test;

/**
 * Translates the "assess" describe block of {@code test/assess.test.ts} 1:1.
 * The {@code makeProviderEvaluator}/{@code SynthesisProvider} describe block in
 * the TS reference has no Java equivalent — this port has no LLM provider
 * abstraction, by design (see {@link Assess} and {@link ConfidenceEvaluator}).
 */
class AssessTest {

    private static final List<GroundUnit> UNITS =
            List.of(new GroundUnit("risk-policy", "sha-risk-policy", "Risk assessments require dual sign-off."));

    private static ConfidenceSignal signal(String source, double score, String reasoning) {
        return new ConfidenceSignal(source, score, reasoning);
    }

    private static ConfidenceEvaluator evaluatorOf(double score, String reasoning) {
        return (task, answer, units) -> signal(ConfidenceSignal.EVALUATOR, score, reasoning);
    }

    @Test
    void passesWhenTheAdjudicatedScoreClearsTheThreshold() {
        ConfidenceVerdict v = Assess.assess("draft risk assessment", "Low risk.", UNITS,
                AssessOptions.builder(0.7)
                        .selfReport(signal(ConfidenceSignal.SELF, 0.9, "clear-cut case"))
                        .build());

        assertEquals("confidence", v.gate());
        assertTrue(v.passed());
        assertEquals(0.9, v.score());
        assertEquals(0.7, v.threshold());
        assertEquals(1, v.signals().size());
        assertTrue(v.detail().contains("0.9"));
        assertTrue(v.detail().contains("0.7"));
    }

    @Test
    void failsWithAWrittenSpecificReasonWhenBelowThreshold() {
        ConfidenceVerdict v = Assess.assess("draft risk assessment", "Unsure.", UNITS,
                AssessOptions.builder(0.7)
                        .severity("critical")
                        .selfReport(signal(ConfidenceSignal.SELF, 0.55, "conflicting inputs"))
                        .build());

        assertFalse(v.passed());
        assertEquals(0.55, v.score());
        assertEquals("critical", v.severity());
        assertTrue(v.detail().contains("0.55"));
        assertTrue(v.detail().contains("conflicting inputs"));
    }

    @Test
    void aggregatesMultipleSignalsWithMinByDefaultFailClosed() {
        ConfidenceVerdict v = Assess.assess("task", "Answer. Confidence: 0.9", UNITS,
                AssessOptions.builder(0.7)
                        .evaluator(evaluatorOf(0.6, "citations are thin"))
                        .build());

        assertEquals(2, v.signals().size());
        assertEquals(0.6, v.score());
        assertFalse(v.passed());
        assertTrue(v.detail().contains("citations are thin"));
    }

    @Test
    void aggregateMeanAveragesTheSignals() {
        ConfidenceVerdict v = Assess.assess("task", "Answer. Confidence: 0.9", UNITS,
                AssessOptions.builder(0.7)
                        .evaluator(evaluatorOf(0.6, "evaluator judgment"))
                        .aggregate("mean")
                        .build());

        assertEquals(0.75, v.score(), 1e-9);
        assertTrue(v.passed());
    }

    @Test
    void noObtainableSignalFailsClosedWithASpecificDetail() {
        ConfidenceVerdict v = Assess.assess("task", "An answer with no self-report.", UNITS,
                AssessOptions.builder(0.7).build());

        assertFalse(v.passed());
        assertEquals(0, v.score());
        assertEquals(0, v.signals().size());
        assertTrue(v.detail().toLowerCase().contains("no confidence signal"));
    }

    @Test
    void evaluatorFailureFailsClosedTheErrorBecomesTheDetail() {
        ConfidenceEvaluator broken = (task, answer, units) -> {
            throw new RuntimeException("provider offline");
        };
        ConfidenceVerdict v = Assess.assess("task", "Answer. Confidence: 0.95", UNITS,
                AssessOptions.builder(0.7).evaluator(broken).build());

        assertFalse(v.passed());
        assertTrue(v.detail().contains("provider offline"));
    }

    @Test
    void evaluatorReturningAnOutOfRangeScoreFailsClosed() {
        ConfidenceVerdict v = Assess.assess("task", "Answer.", UNITS,
                AssessOptions.builder(0.7).evaluator(evaluatorOf(42, "not calibrated")).build());

        assertFalse(v.passed());
        String detail = v.detail().toLowerCase();
        assertTrue(detail.contains("out of range") || detail.contains("invalid"));
    }

    @Test
    void rawSignalsArePreservedVerbatimForCalibration() {
        ConfidenceVerdict v = Assess.assess("task", "Answer. Confidence: 80%", UNITS,
                AssessOptions.builder(0.5).evaluator(evaluatorOf(0.9, "well grounded")).build());

        List<String> sources = v.signals().stream().map(ConfidenceSignal::source).sorted().toList();
        assertEquals(List.of(ConfidenceSignal.EVALUATOR, ConfidenceSignal.SELF), sources);
        for (ConfidenceSignal s : v.signals()) {
            assertTrue(s.reasoning() != null && !s.reasoning().isBlank());
        }
    }

    @Test
    void includeSelfReportFalseIgnoresTheAnswersSelfReport() {
        ConfidenceVerdict v = Assess.assess("task", "Answer. Confidence: 0.2", UNITS,
                AssessOptions.builder(0.7)
                        .evaluator(evaluatorOf(0.9, "evaluator judgment"))
                        .includeSelfReport(false)
                        .build());

        assertEquals(1, v.signals().size());
        assertEquals(0.9, v.score());
        assertTrue(v.passed());
    }

    @Test
    void stampsAsOfCallerProvidedWinsForReproducibility() {
        ConfidenceVerdict v = Assess.assess("task", "Answer.", UNITS,
                AssessOptions.builder(0.7)
                        .selfReport(signal(ConfidenceSignal.SELF, 0.8, "because"))
                        .asOf("2026-07-20")
                        .build());

        assertEquals("2026-07-20", v.asOf());
    }

    @Test
    void rejectsAnInvalidThresholdCallerErrorNotAVerdict() {
        IllegalArgumentException tooHigh = assertThrows(IllegalArgumentException.class,
                () -> Assess.assess("task", "Answer.", UNITS, AssessOptions.builder(7).build()));
        assertTrue(tooHigh.getMessage().contains("threshold"));

        IllegalArgumentException tooLow = assertThrows(IllegalArgumentException.class,
                () -> Assess.assess("task", "Answer.", UNITS, AssessOptions.builder(-0.1).build()));
        assertTrue(tooLow.getMessage().contains("threshold"));
    }

    // --- extractSelfReport (test/assess.test.ts, describe("extractSelfReport")) ---

    @Test
    void extractSelfReportParsesADecimalConfidenceLine() {
        ConfidenceSignal s = Assess.extractSelfReport("The risk is low.\n\nConfidence: 0.82");
        assertEquals(0.82, s.score(), 1e-9);
        assertEquals(ConfidenceSignal.SELF, s.source());
    }

    @Test
    void extractSelfReportParsesAPercentage() {
        assertEquals(0.82, Assess.extractSelfReport("Confidence: 82%").score(), 1e-9);
    }

    @Test
    void extractSelfReportTakesTheLastReportWhenSeveralAppear() {
        ConfidenceSignal s = Assess.extractSelfReport("Confidence: 0.9 early on.\nFinal confidence: 0.6");
        assertEquals(0.6, s.score(), 1e-9);
    }

    @Test
    void extractSelfReportReturnsNullWhenNoSelfReportExists() {
        assertNull(Assess.extractSelfReport("Just an answer."));
    }

    @Test
    void extractSelfReportClampsRunawayPercentagesIntoRange() {
        assertEquals(1.0, Assess.extractSelfReport("Confidence: 150%").score());
    }
}
